src/Microsoft.SqlTools.ServiceLayer/Utility/SqlScriptFormatters/FromSqlScript.cs

```diff
@@ -0,0 +1,148 @@
+﻿//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Text.RegularExpressions;
+using Microsoft.SqlTools.Utility;
+
+namespace Microsoft.SqlTools.ServiceLayer.Utility.SqlScriptFormatters
+{
+    /// <summary>
+    /// Provides utilities for converting from SQL script syntax into POCOs.
+    /// </summary>
+    public static class FromSqlScript
+    {
+        private static readonly Regex StringRegex = new Regex("^N?'(.*)'$", RegexOptions.Compiled);
+        
+        /// <summary>
+        /// Decodes a multipart identifier as used in a SQL script into an array of the multiple
+        /// parts of the identifier. Implemented as a state machine that iterates over the
+        /// characters of the multipart identifier.
+        /// </summary>
+        /// <param name="multipartIdentifier">Multipart identifier to decode (eg, "[dbo].[test]")</param>
+        /// <returns>The parts of the multipart identifier in an array (eg, "dbo", "test")</returns>
+        /// <exception cref="FormatException">
+        /// Thrown if an invalid state transition is made, indicating that the multipart identifer
+        /// is not valid. 
+        /// </exception>
+        public static string[] DecodeMultipartIdenfitier(string multipartIdentifier)
+        {
+            StringBuilder sb = new StringBuilder();
+            List<string> namedParts = new List<string>();
+            bool insideBrackets = false;
+            bool bracketsClosed = false;
+            for (int i = 0; i < multipartIdentifier.Length; i++)
+            {
+                char iChar = multipartIdentifier[i];
+                if (insideBrackets)
+                {
+                    if (iChar == ']')
```

Comment:
Are there cases where we'd have quoted identifiers `"dbo"."test"` instead of brackets? Do we need to account for that? Or do we always control the input here? If this is used for arbitrary scripts this is important. If it's just for internally produced ones that always have `[` then it can be ignored

Reply:
Right, this is only used for input from ourselves. The values that are being processed here are from SMO or SqlClient. I'm not 100% certain on the SMO stuff (only stuff I'm using is default constraint values) because the values aren't escaped already.